lib/livebook/session.ex

```diff
@@ -991,17 +992,38 @@ defmodule Livebook.Session do
   end
 
   defp maybe_save_notebook_sync(state) do
-    if should_save_notebook?(state) do
+    file = notebook_autosave_file(state)
+
+    if file && should_save_notebook?(state) do
       content = LiveMarkdown.Export.notebook_to_markdown(state.data.notebook)
-      result = FileSystem.File.write(state.data.file, content)
+      result = FileSystem.File.write(file, content)
       handle_save_finished(state, result)
     else
       state
     end
   end
 
   defp should_save_notebook?(state) do
-    state.data.file != nil and state.data.dirty and state.save_task_pid == nil
+    state.data.dirty and state.save_task_pid == nil
+  end
+
+  defp notebook_autosave_file(state) do
+    state.data.file || default_notebook_file(state)
+  end
+
+  defp default_notebook_file(session) do
+    if path = Livebook.Config.autosave_path() do
+      dir = path |> FileSystem.Utils.ensure_dir_path() |> FileSystem.File.local()
+      filename = name_with_timestamp(session.created_at) <> ".livemd"
```

Comment:
I would do `#{date}/#{time}-#{random_number}.livemd`. The directory for the date is to make things a bit more manageable?

Reply:
Sounds good! Anything specific for `random_number`? We could just use the milliseconds part?

Reply:
I am worried miliseconds is not random enough. Especially on a VM that can do a bunch of stuff submilisecond. :D `:crypto.strong_rand_bytes(3) |> Base.url_encode64()` would do it for me.

Reply:
The time is based on session creation, so basically there would need to be two sessions created at the same millisecond, which is probably impossible, since they are spawned by a single supervisor ^^ But a random part sounds good too!

Reply:
But then that brings another question: are we blocking the supervisor while creating sessions? Shouldn't this initialization, like creating directories, happen in a `handle_continue` callback instead?

Reply:
Nono, the directory is created only when persisting. My point is that we take the timestamp when initializing the session and use it later. But since the filename has so many numbers, I like some random characters more :)

Reply:
Ah, got it!

Reply:
I went with `String.slice(session_id, 0..3)`, since we want the random pat to be deterministic and session id gives us that for free!